swarm_dial.go

```diff
@@ -104,75 +118,81 @@ func (db *dialbackoff) Backoff(p peer.ID) (backoff bool) {
 	return false
 }
 
-const baseBackoffTime = time.Second * 5
-const maxBackoffTime = time.Minute * 5
+// BackoffBase is the base amount of time to backoff (default: 5s).
+var BackoffBase = time.Second * 5
+
+// BackoffCoef is the backoff coefficient (default: 1s).
+var BackoffCoef = time.Second
+
+// BackoffMax is the maximum backoff time (default: 5m).
+var BackoffMax = time.Minute * 5
 
 // AddBackoff lets other nodes know that we've entered backoff with
 // peer p, so dialers should not wait unnecessarily. We still will
 // attempt to dial with one goroutine, in case we get through.
-func (db *dialbackoff) AddBackoff(p peer.ID) {
+//
+// Backoff is not exponential, it's quadratic and computed according to the
+// following formula:
+//
+//     BackoffBase + BakoffCoef * PriorBackoffs^2
+//
+// Where PriorBackoffs is the number of previous backoffs.
+func (db *DialBackoff) AddBackoff(p peer.ID) {
 	db.lock.Lock()
 	defer db.lock.Unlock()
 	db.init()
 	bp, ok := db.entries[p]
 	if !ok {
 		db.entries[p] = &backoffPeer{
 			tries: 1,
-			until: time.Now().Add(baseBackoffTime),
+			until: time.Now().Add(BackoffBase),
 		}
 		return
 	}
 
-	expTimeAdd := time.Second * time.Duration(bp.tries*bp.tries)
-	if expTimeAdd > maxBackoffTime {
-		expTimeAdd = maxBackoffTime
+	backoffTime := BackoffBase + BackoffCoef*time.Duration(bp.tries*bp.tries)
```

Comment:
Nit: Adding in the BackoffBase *here* ensures that we don't go over the maximum (we used to go over by BackoffBase).